hazelcast/src/main/java/com/hazelcast/instance/HazelcastInstanceFactory.java

```diff
@@ -18,12 +18,14 @@
 
 import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.spi.annotation.PrivateApi;
 
 /***
  * This is interface which provides capability for Hazelcast factories customization;
  * DefaultHazelcastInstanceFactory is used by default, but external service can have;
  * custom implementation which should be specified in META-INF services;
  */
+@PrivateApi
```

Comment:
`NodeExtension` is private. 
It has never been meant to be an extension point. I saw external projects use it for 2 use-cases:
1. to hack in a custom joiner impl. this is no longer needed with Discovery SPI.
2. to hack off-heap store in.